blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md

```diff
@@ -0,0 +1,238 @@
+    Title: Bullets are good for your Coq proofs
+    Date: 2017-02-21T19:04:28
+    Tags: coq, by Gabriel Scherer
+
+I believe that bullets are one of the most impactful features of
+recent versions of Coq, among those that non-super-expert users can
+enjoy. They had a big impact on the maintainability of my
+proofs. Unfortunately, they are not very well-known, due to the fact
+that some introductory documents have not been updated to use them.
+
+Bullets are a very general construction and there are several possible
+ways to use them; I have iterated through different styles. In this
+post I will give the general rules, and explain my current usage
+style.
+
+<!-- more -->
+
+# Why bullets
+
+While you are doing a proof, Coq shows a list of subgoals that have to
+be proved before the whole proof is complete. Most proof steps will
+operate on the current active subgoal, changing the hypotheses or the
+goal to prove, but some proof steps will split it into several
+subgoals (growing the total list of goals), or may terminate the proof
+of the current subgoal and show you the next active subgoal.
+
+Before bullets, a typical proof script would contain the proofs of
+each subgoal, one after another.
+
+```
+induction foo. (* this creates many subgoal *)
+
+proof of first subgoal.
+
+proof of second subgoal.
+```
+
+There are many ways to structure this to make the structure more
+apparent: people would typically have a comment on each subgoal, or
+make disciplined use of indentation and blank lines. But, in my
+experience, a major problem with this style was maintenability in the
+face of changes to the definitions or parts of automation. It could be
+very hard of what was happening when a proof suddenly broke after
+a change before in the file:
+
+- If a proof step now proves *less* things, then what used to be the
+  end of a subgoal may not be anymore. Coq would then start reading
+  the proof of the next subgoal and try to apply it to the unfinished
+  previous goals, generating very confusing errors (you believe you
+  are in the second subgoal, but the context talks about a leaf case
+  of the first goal).
+
+- If a proof step now proves *more* things, it is also very bad! The
+  next proof steps, meant for the first subgoal (for example), would
+  then apply to the beginning of the second subgoal, and you get very
+  confusing errors again.
+
+What we need for robustness is a way to indicate our *intent* to Coq,
+when we think that a subgoal is finished and that a new subgoal
+starts, so that Coq can fail loudly at the moment where it notices
+that this intent does not match reality, instead of at an arbitrary
+later time.
+
+(The `S*Case` tactics used in Software Foundations can solve this
```

Comment:
SF: https://www.cis.upenn.edu/~bcpierce/sf/

I tried to find a link to S*Case from SF, but it seems they recommend bullets now. Yay!
https://www.cis.upenn.edu/~bcpierce/sf/current/Basics.html